### PR TITLE
Improvements when calling the crontab cmd.

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
Modified code to make sure the variable "$cronListCmd" is always evaluated as a command and not simply referenced as a quoted string.

### CODE REVIEW

**Comment by ExtremeFiretop** [comment]
I'll approve and merge this, but I honestly have little idea how you got here lol.
Like dark magic to me. Maybe we can continue the discussion here, the other thread is getting long.

My last response of value there was:

> > Maybe we try making cru l the default first attempt instead of crontab?
> > ```
> > if [ -z "$(which cru)" ]; then
> >     cronListCmd="crontab -l"
> > else
> >     cronListCmd="cru l"
> > fi
> > ```
> 
> The problem appears to be how the variable "$cronListCmd" is evaluated when referenced. Apparently, in some router environments, sometimes the variable is **not** evaluated as containing a command but as a referenced quoted string. So I changed each call to make sure each time it's forced to be evaluated as a command (see PR #176).
> 
> Since I could not recreate the problem on my router, please try to check if the latest changes worked in your router that experienced the problem.
> 
> Thanks.

I have zero idea how you got there.
I see the error: **printf "${REDct}**ERROR**${NOct}: Failed to add the cron job [${CRON_JOB_TAG}].\n"**
I look in the code for that error and find it in this **if statement:**

```
       if _AddCronJobEntry_
       then
           printf "Cron job '${GRNct}${CRON_JOB_TAG}${NOct}' was added successfully.\n"
           _AddCronJobRunScriptHook_
       else
           printf "${REDct}**ERROR**${NOct}: Failed to add the cron job [${CRON_JOB_TAG}].\n"
       fi
```

I review the **_AddCronJobEntry_** function that called it.
Inside the function, I clearly see:

```
   cru a "$CRON_JOB_TAG" "$newSchedule $CRON_JOB_RUN"
   sleep 1
   if $cronListCmd | grep -qE "$CRON_JOB_RUN #${CRON_JOB_TAG}#$"
```

So it calls the cronListCmd commandlet we created, looking up the cronListCmd I see command for crontab -l:

```
if [ -z "$(which cru)" ]; then
    cronListCmd="crontab -l"
else
    cronListCmd="cru l"
fi
```

Which is why I figured you made the change you did yesterday to that format, from the old format:

```
cronCmd="$(which crontab) -l"
[ "$cronCmd" = " -l" ] && cronCmd="$(which cru) l"
```

How you took away that this line is the problem:

```
    if [ "$1" = "**ERROR**" ] && [ "$2" = "**NO_URL**" ]
    then
        Say "${REDct}**ERROR**${NOct}: No firmware release URL was found for [$PRODUCT_ID] router model."
        "$inMenuMode" && _WaitForEnterKey_ "$mainMenuReturnPromptStr"
        return 1
    fi
```

I have zero idea, I would assume to see 
"**${REDct}**ERROR**${NOct}: No firmware release URL was found for [$PRODUCT_ID] router model.""**

Considering that's the if statement for if any of those aren't populated.
My default thought would be that 

```
    # Use set to read the output of the function into variables
    set -- $(_GetLatestFWUpdateVersionFromWebsite_ "$FW_URL_RELEASE")
    release_version="$1"
    release_link="$2"
```

Is working fine normally, since we use release_version="$1" right below it in: 

```
    if ! _CheckTimeToUpdateFirmware_ "$current_version" "$release_version"
    then
        "$inMenuMode" && _WaitForEnterKey_ "$mainMenuReturnPromptStr"
        return 0
    fi
```

And use release_link="$2" right here:

```
    # Get the required space for the firmware download and extraction
    required_space_kb=$(get_required_space "$release_link")
    if ! _HasRouterMoreThan256MBtotalRAM_ && [ "$required_space_kb" -gt 51200 ]; then
        if ! _ValidateUSBMountPoint_ "$FW_ZIP_BASE_DIR" 1
        then
            Say "${REDct}**ERROR**${NOct}: A USB drive is required for the F/W update due to limited RAM."
            "$inMenuMode" && _WaitForEnterKey_ "$mainMenuReturnPromptStr"
            return 1
        fi
    fi
```

I'd think one of those would complain if we had a problem with any of those values, right?
Maybe you can explain your train of thought, troubleshooting, etc so I can learn and take away how the heck you made this connection?

File: MerlinAU.sh [hunk 1]
```diff
@@ -2419,7 +2419,7 @@ _AddCronJobEntry_()
 
    cru a "$CRON_JOB_TAG" "$newSchedule $CRON_JOB_RUN"
    sleep 1
-   if $cronListCmd | grep -qE "$CRON_JOB_RUN #${CRON_JOB_TAG}#$"
+   if eval $cronListCmd | grep -qE "$CRON_JOB_RUN #${CRON_JOB_TAG}#$"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2422]: Approved.

File: MerlinAU.sh [hunk 2]
```diff
@@ -2434,10 +2434,10 @@ _AddCronJobEntry_()
 _DelCronJobEntry_()
 {
    local retCode
-   if $cronListCmd | grep -qE "$CRON_JOB_RUN #${CRON_JOB_TAG}#$"
+   if eval $cronListCmd | grep -qE "$CRON_JOB_RUN #${CRON_JOB_TAG}#$"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2437]: Approved.

File: MerlinAU.sh [hunk 3]
```diff
@@ -3053,19 +3060,13 @@ Please manually update to version $minimum_supported_version or higher to use th
...
-    fi
-
     ##---------------------------------------##
     ## Added by ExtremeFiretop [2023-Dec-09] ##
     ##---------------------------------------##
     # Get the required space for the firmware download and extraction
-    required_space_kb=$(get_required_space "$release_link")
-    if ! _HasRouterMoreThan256MBtotalRAM_ && [ "$required_space_kb" -gt 51200 ]; then
+    required_space_kb="$(get_required_space "$release_link")"
+    if ! _HasRouterMoreThan256MBtotalRAM_ && [ "$required_space_kb" -gt 51200 ]
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:3068]: Approved.

File: MerlinAU.sh [hunk 4]
```diff
@@ -3919,7 +3920,7 @@ if [ "$FW_UpdateCheckState" -eq 1 ]
 then
     runfwUpdateCheck=true
     # Check if the CRON job already exists #
-    if ! $cronListCmd | grep -qE "$CRON_JOB_RUN #${CRON_JOB_TAG}#$"
+    if ! eval $cronListCmd | grep -qE "$CRON_JOB_RUN #${CRON_JOB_TAG}#$"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:3923]: Approved.

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

 And to clarify, I understand the if statement didn't give us the message: 

""${REDct}ERROR${NOct}: No firmware release URL was found for [$PRODUCT_ID] router model.""

Because the values are populated, and they just aren't being evaluated as an executable.

But like... How did you come up to this conclusion is the part that baffles me. That connection is the part that feels like dark magic.

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

WRT to the **crontab** error, the earlier error message reported by the forum user was this:
`/usr/bin/crontab -l: not found`

Then yesterday you reported this one:
`crontab -l: not found`

Putting 2 & 2 together plus some intuition (based on past experience), I concluded that the problem was most likely due to how the variable definition was being evaluated in some instances (I still don't understand why the error happens in some cases but the same code works most of the time).

Here's a function that I think illustrates the problem better than explaining it in words:

```
_TEST_CODE1_()
{
   local testCmd="crontab -l"
   
   printf "\nTEST#1 using [\$testCmd]:\n"
   $testCmd | head -n 5

   printf "\nTEST#2 using [eval \$testCmd]:\n"
   eval $testCmd | head -n 5

   printf "\nTEST#3 using [eval \"\$testCmd\"]:\n"
   eval "$testCmd" | head -n 5

   printf "\nTEST#4 using [\"\$testCmd\"]:\n"
   "$testCmd" | head -n 5
}
```

Running the above function, you'll see the last test statement recreates the same error message where the variable is not evaluated as a "command" but as a "quoted string" variable. So the fix is to force the call to always evaluate it as a command.

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

The other error message you reported was this one:
`/jffs/scripts/MerlinAU.sh: line 17: 2: parameter not set`

This indicates that the error was coming from using the unset variable "`$2`" from within the "`_RunFirmwareUpdateNow_`" function, and based on where it happened I figured the most likely cause was this line:
`if [ "$1" = "**ERROR**" ] && [ "$2" = "**NO_URL**" ]`

I could not recreate the problem at all on my router, but here's another sample code that demonstrates how it can happen:

```
_TEST_CODE2_()
{
   local param1  param2
   set -u
   printf "\nTEST #1:\n"
   set -- $(echo "Hello" ; echo "World")
   param1="$1"
   param2="$2"
   printf "Returned Param#1=[$1]\n"
   printf "Returned Param#2=[$2]\n"
   
   printf "\nTEST #2:\n"
   set -- $(echo "Hello" ; echo "")
   param1="$1"
   param2="$2"
   printf "Returned Param#1=[$1]\n"
   printf "Returned Param#2=[$2]\n"
}
```

When running the above function, you'll see the same error message. But again, since I can't recreate the problem, all I can do right now is improve the code to avoid running into the same situation.

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop,
> 
> The other error message you reported was this one: `/jffs/scripts/MerlinAU.sh: line 17: 2: parameter not set`
> 
> This indicates that the error was coming from using the unset variable "`$2`" from within the "`_RunFirmwareUpdateNow_`" function, and based on where it happened I figured the most likely cause was this line: `if [ "$1" = "**ERROR**" ] && [ "$2" = "**NO_URL**" ]`
> 
> I could not recreate the problem at all on my router, but here's another sample code that demonstrates how it can happen:
> 
> ```
> _TEST_CODE2_()
> {
>    local param1  param2
>    set -u
>    printf "\nTEST #1:\n"
>    set -- $(echo "Hello" ; echo "World")
>    param1="$1"
>    param2="$2"
>    printf "Returned Param#1=[$1]\n"
>    printf "Returned Param#2=[$2]\n"
>    
>    printf "\nTEST #2:\n"
>    set -- $(echo "Hello" ; echo "")
>    param1="$1"
>    param2="$2"
>    printf "Returned Param#1=[$1]\n"
>    printf "Returned Param#2=[$2]\n"
> }
> ```
> 
> When running the above function, you'll see the same error message. But again, since I can't recreate the problem, all I can do right now is improve the code to avoid running into the same situation.

So that's exactly it, when I noticed the issues (both of them as I understand it) I instantly was unable to recreate it again in both routers. So I wasn't sure how to go about troubleshooting it. When I reported them I was hoping you were gonna say you also noticed it but we're unable to recreate it afterwards (like me). But inside you don't have the issue at all and you still came up with a theory and solution which is how I was just mind blown.

I understand the proposed solution and the theory on the issue, but I also don't know why I was instantly unable to recreate the issue again.

I'll test these functions tomorrow and report back! Thank you for fixing the missed variable in the release for GitHub too btw!

**Comment by ExtremeFiretop** [comment]
Goodnight buddy!

**Comment by ExtremeFiretop** [comment]
> _TEST_CODE2_()
> {
>    local param1  param2
>    set -u
>    printf "\nTEST #1:\n"
>    set -- $(echo "Hello" ; echo "World")
>    param1="$1"
>    param2="$2"
>    printf "Returned Param#1=[$1]\n"
>    printf "Returned Param#2=[$2]\n"
>    
>    printf "\nTEST #2:\n"
>    set -- $(echo "Hello" ; echo "")
>    param1="$1"
>    param2="$2"
>    printf "Returned Param#1=[$1]\n"
>    printf "Returned Param#2=[$2]\n"
> }

Admin@GT-AXE11000-DBF0:/tmp/home/root# _TEST_CODE2_()
> {
>    local param1  param2
>    set -u
>    printf "\nTEST #1:\n"
>    set -- $(echo "Hello" ; echo "World")
>    param1="$1"
>    param2="$2"
>    printf "Returned Param#1=[$1]\n"
>    printf "Returned Param#2=[$2]\n"
>
>    printf "\nTEST #2:\n"
>    set -- $(echo "Hello" ; echo "")
>    param1="$1"
>    param2="$2"
>    printf "Returned Param#1=[$1]\n"
>    printf "Returned Param#2=[$2]\n"
> }

Admin@GT-AXE11000-DBF0:/tmp/home/root# _TEST_CODE2_

TEST #1:
Returned Param#1=[Hello]
Returned Param#2=[World]

TEST #2:
-sh: 2: parameter not set

**Comment by ExtremeFiretop** [comment]
Admin@GT-AXE11000-DBF0:/tmp/home/root# _TEST_CODE1_()
> {
>    local testCmd="crontab -l"
>
>    printf "\nTEST#1 using [\$testCmd]:\n"
>    $testCmd | head -n 5
>
>    printf "\nTEST#2 using [eval \$testCmd]:\n"
>    eval $testCmd | head -n 5
>
>    printf "\nTEST#3 using [eval \"\$testCmd\"]:\n"
>    eval "$testCmd" | head -n 5
>
>    printf "\nTEST#4 using [\"\$testCmd\"]:\n"
>    "$testCmd" | head -n 5
> }
Admin@GT-AXE11000-DBF0:/tmp/home/root#
Admin@GT-AXE11000-DBF0:/tmp/home/root# _TEST_CODE1_

TEST#1 using [$testCmd]:
45 */6 * * * /jffs/addons/amtm/routerdate cron #amtm_RouterDate#
45 2 * * * sh /jffs/scripts/backupmon.sh #RunBackupMon#
10 7 * * Sun /bin/sh /jffs/addons/amtm/sc_update.mod -run #amtm_ScriptsUpdateNotification#
0 4 * * * sh /jffs/scripts/MerlinAU.sh run_now #MerlinAU#

TEST#2 using [eval $testCmd]:
45 */6 * * * /jffs/addons/amtm/routerdate cron #amtm_RouterDate#
45 2 * * * sh /jffs/scripts/backupmon.sh #RunBackupMon#
10 7 * * Sun /bin/sh /jffs/addons/amtm/sc_update.mod -run #amtm_ScriptsUpdateNotification#
0 4 * * * sh /jffs/scripts/MerlinAU.sh run_now #MerlinAU#

TEST#3 using [eval "$testCmd"]:
45 */6 * * * /jffs/addons/amtm/routerdate cron #amtm_RouterDate#
45 2 * * * sh /jffs/scripts/backupmon.sh #RunBackupMon#
10 7 * * Sun /bin/sh /jffs/addons/amtm/sc_update.mod -run #amtm_ScriptsUpdateNotification#
0 4 * * * sh /jffs/scripts/MerlinAU.sh run_now #MerlinAU#

TEST#4 using ["$testCmd"]:
-sh: crontab -l: not found

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

Absolutely amazing catch. 
This makes me feel like I can kinda rest on those bugs now.

**Comment by ExtremeFiretop** [comment]
We may want to make a release at some point just to patch those up, and to test the latest changes with our release flow :)

**Comment by Martinski4GitHub** [comment]
> We may want to make a release at some point just to patch those up, and to test the latest changes with our release flow :)
>
Yeah, I think we will need a patch release fairly soon, especially since we don't want to have those kinds of weird, unexplainable errors "out in the wild" for too long. They appear to be rare but it's better to "nip it in the bud."

**Comment by ExtremeFiretop** [comment]
> > We may want to make a release at some point just to patch those up, and to test the latest changes with our release flow :)
> 
> Yeah, I think we will need a patch release fairly soon, especially since we don't want to have those kinds of weird, unexplainable errors "out in the wild" for too long. They appear to be rare but it's better to "nip it in the bud."

Agreed. I think we roll out 1.0.10 soon.

My PR can be for a 1.1.0 release if you want.
I'm going to ask Viktor to see if he can try it out with his nodes and just report back the general situation.

**Comment by Martinski4GitHub** [comment]
> We may want to make a release at some point just to patch those up, and to test the latest changes with our release flow :)
>
This evening, I was thinking of adding some code to the "**check_memory_and_prompt_reboot**" function to gather some debug data WRT memory stats at the points where the code is trying to free up memory more aggressively. I believe that having that extra data in the log file might give us some clues about why some routers cannot get memory back after the contents of the ZIP file have been extracted.

What do you think?  It's fairly simple & quick to do it.

**Comment by ExtremeFiretop** [comment]
> > We may want to make a release at some point just to patch those up, and to test the latest changes with our release flow :)
> 
> This evening, I was thinking of adding some code to the "**check_memory_and_prompt_reboot**" function to gather some debug data WRT memory stats at the points where the code is trying to free up memory more aggressively. I believe that having that extra data in the log file might give us some clues about why some routers cannot get memory back after the contents of the ZIP file have been extracted.
> 
> What do you think? It's fairly simple & quick to do it.

Honestly at this point we need as much data as we can get because the answer is not clear cut. It's a black box and the behavior isn't self explanatory what is going on .

If you want to pump all the data into that last step of the cleanup function, your more than welcome, be my guest, not even, make yourself at home. Pull up a chair and rip the function a new one 😂😜

**Comment by Martinski4GitHub** [comment]
> > Yeah, I think we will need a patch release fairly soon, especially since we don't want to have those kinds of weird, unexplainable errors "out in the wild" for too long. They appear to be rare but it's better to "nip it in the bud."
> 
> Agreed. I think we roll out 1.0.10 soon.
> 
> My PR can be for a 1.1.0 release if you want. I'm going to ask Viktor to see if he can try it out with his nodes and just report back the general situation.
>
Yeah, we need a 2nd pair of eyes to test & verify the new code. Perhaps this also would be a good time to create a **Beta** release given the new functionality so that those users with AiMesh nodes can test it and report any problems before the code goes to production. Just a thought.

**Comment by Martinski4GitHub** [comment]
> > What do you think? It's fairly simple & quick to do it.
> 
> Honestly at this point we need as much data as we can get because the answer is not clear cut. It's a black box and the behavior isn't self explanatory what is going on .
> 
> If you want to pump all the data into that last step of the cleanup function, your more than welcome, be my guest, not even, make yourself at home. Pull up a chair and rip the function a new one 😂😜
>
I'm thinking of adding the code **before** & **after** the sections where the function is trying to free memory so we can have the "BEFORE" & "AFTER" snapshots to see what changed & what didn't. I'll add the code so you'll be able to see for yourself what I mean.

**Pull request merged**